### ISSUE TITLE
[Feature] use emojis as a simple solution for icons

### ISSUE THREAD

**Issue by eirinnm** [jarun/nnn#1346]
**Is your feature request related to a problem? Please describe.**

Problem: getting icons in nnn requires jumping through hoops with third party packages: either patching your system font with nerdfont, or using icons-in-terminal, which appears to be abandoned (no updates in 5 years, PRs getting ignored).

**Describe the solution you'd like**

nnn should have an option to use emojis as icons. Emojis are widely supported now. For instance, ubuntu 18.04 supports colour emojis. 

Obviously, we can't have unique icons for every filetype. But here are some examples, which would make nnn really nice to use:

📄 file  
📂 folder  
📑 executable file  
🔗 symbolic link
📦️ archive

Since emojis are just unicode codepoints and the actual artwork is provided by the operating system, this feature should not cause any bloat.

**Comment by N-R-K** [comment]
> Emojis are widely supported now. 

Untrue.

1. Many DIY distros don't include emoji font by default
2. Any terminal using XFT will either crash trying to display color emojies or will ignore color emoji font in order to avoid crashing. https://git.suckless.org/st/file/FAQ.html#l224
3. I don't believe (most ?) bitmap fonts, which are quite common among terminal users, support emojies.

> Since emojis are just unicode codepoints and the actual artwork is provided by the operating system, this feature should not cause any bloat.

Depends on the font. `noto-emoji` is 165Mb for example. Just because it comes preinstalled on some systems doesn't make it non bloated.

**Comment by MuhammedZakir** [comment]
Not specfically related to nnn: if you want to avoid patching or other headaches, you can use a terminal that support user-configured font fallback option, such as WezTerm and foot. This makes it really easy to use your favorite font along with, for example, nerd icons.

**Comment by cainnz** [comment]
thank you for all of you guys response, @MuhammedZakir I will go ahead and try one of those terminals. I really love nnn! thank you for all you guys do.

**Comment by eirinnm** [comment]
>Do you see the contradiction that arises from these two statements? Yes, this way or that your system font (or rather the one your terminal uses) has to support displaying the desired symbol. 

I don't see the contradiction; in one case, I have to patch my fonts, in the other case, it's already been done. 

My specific use-case and reason for suggesting this feature is that I want to roll out nnn to a number of Ubuntu 20.04 workstations via ansible. Installing nnn via ansible is straightforward and easily justified. Installing custom fonts to get the icons is not. The icons-in-terminal repository hasn't been updated in 5 years; naturally I am reluctant to incorporate that into my ansible playbook. But it would still be nice for my end users if nnn could have attractive icons. A built-in optional feature to display emojis as icons would be a suitable middle ground between no icons and the full icons-in-terminal installation.

>Many DIY distros don't include emoji font by default

Understood. This option is intended for users on OSes where this is already supported, like Gnome in Ubuntu 18.04 and higher. Ideally a terminal should gracefully handle unicode symbols that it can't render.

**Comment by N-R-K** [comment]
> Ideally a terminal should gracefully handle unicode symbols that it can't render.

Ideally sure, but correctly handling unicode is a much more complex topic. Especially because a single code-point isn't a always a single user perceived character.

In your very post, the archive emoji is actually 2 different code points which make up the "grapheme cluster".

Neither XTerm nor ST seems to be able to deal with it.

![image](https://user-images.githubusercontent.com/79544946/164326105-f00dbd02-c1e9-4ccc-a24d-0826ba129259.png)

So if we *do* add emoji icons, we'll probably need to ensure all the emojies are single code-point or add a disclaimer about it.

**Comment by luukvbaal** [comment]
I suppose a make option `O_EMOJI` could be introduced with a minimal set of icons by filetype instead of by extension.

**Comment by jarun** [comment]
Do these icons work with gnome-terminal, konsole and xfce4-terminal without any issues?

**Comment by jarun** [comment]
@luukvbaal can you please explain "by filetype instead of by extension"? Why so? If an extension doesn't have an icon it remains blank, right?

**Comment by KlzXS** [comment]
There are fallbacks for directories, normal files and executable file in case we don't match any extension or predefined name.

**Comment by luukvbaal** [comment]
> There are fallbacks for directories, normal files and executable file in case we don't match any extension or predefined name

Yeah, this is what I meant. Something like the list posted by OP. 

The assumption being that the emoji set will not have enough icons to choose from to properly represent all of the extensions currently matched by nerdfont/icons-in-terminal with the same granularity.
Although I suppose some extension matching would be desired for archives/pictures/documents etc.

**Comment by jarun** [comment]
Yes, I do not expect any logic change other than whatever we already have. Just the fixed unicodes to be added for a minimal set. Rest the maintainer would have to add.

**Comment by jarun** [comment]
@eirinnm Do these icons work with gnome-terminal, konsole and xfce4-terminal without any issues?

**Comment by jarun** [comment]
I would suggest you document the behaviour for users.

**Comment by luukvbaal** [comment]
I'm not really interested in constructing the header file but here is a list of example emojis for reference:
```
di=📁:\
fi=📃:\
tw=🤝:\
ow=📂:\
ln=⛓:\
or=❌:\
ex=🎯:\
*.txt=✍:\
*.mom=✍:\
*.me=✍:\
*.ms=✍:\
*.png=🖼:\
*.webp=🖼:\
*.ico=🖼:\
*.jpg=📸:\
*.jpe=📸:\
*.jpeg=📸:\
*.gif=🖼:\
*.svg=🗺:\
*.tif=🖼:\
*.tiff=🖼:\
*.xcf=🖌:\
*.html=🌎:\
*.xml=📰:\
*.gpg=🔒:\
*.css=🎨:\
*.pdf=📚:\
*.djvu=📚:\
*.epub=📚:\
*.csv=📓:\
*.xlsx=📓:\
*.tex=📜:\
*.md=📘:\
*.r=📊:\
*.R=📊:\
*.rmd=📊:\
*.Rmd=📊:\
*.m=📊:\
*.mp3=🎵:\
*.opus=🎵:\
*.ogg=🎵:\
*.m4a=🎵:\
*.flac=🎼:\
*.wav=🎼:\
*.mkv=🎥:\
*.mp4=🎥:\
*.webm=🎥:\
*.mpeg=🎥:\
*.avi=🎥:\
*.mov=🎥:\
*.mpg=🎥:\
*.wmv=🎥:\
*.m4b=🎥:\
*.flv=🎥:\
*.zip=📦:\
*.rar=📦:\
*.7z=📦:\
*.tar.gz=📦:\
*.z64=🎮:\
*.v64=🎮:\
*.n64=🎮:\
*.gba=🎮:\
*.nes=🎮:\
*.gdi=🎮:\
*.1=ℹ:\
*.nfo=ℹ:\
*.info=ℹ:\
*.log=📙:\
*.iso=📀:\
*.img=📀:\
*.bib=🎓:\
*.ged=👪:\
*.part=💔:\
*.torrent=🔽:\
*.jar=♨:\
*.java=♨:\
```

**Comment by jarun** [comment]
@N-R-K or @eirinnm would you like to try?

**Comment by jarun** [comment]
@luukvbaal what are the values to add in the header?
@eirinnm what is the emoji package name for xubuntu?

**Comment by dnkl** [comment]
Some of these have a `wcwidth()` of 1, while the glyph itself is often double-width:

⛓
✍
🖼
🗺
🖌
♨

I would avoid these, as you are likely to see different behavior depending on both the terminal that renders them, but also on the exact font(s) that are installed on the system (and obviously, how font resolving has been configured).

All the glyphs above are emojis with a default _presentation style_ of "text" (which is why their wcwidth is 1). If you want the emoji-style version of it, you can add a `\U+FE0F` character after it. This will enforce "emoji" presentation. But not all terminals respect presentation style selectors, so again, result will be different in different terminal emulators.

Example (from foot): `echo -e 'text presentation: ✍   emoji presentation: ✍\ufe0f'`
![grim-8143d8ba-c31e-11ec-9957-3c9180265d07](https://user-images.githubusercontent.com/14031526/164913767-4479cd35-6e08-42aa-9a8b-90713a747c87.png)

**Comment by jarun** [comment]
Please refer to `icons-in-terminal.h` file. We use the same method for nerd also.

**Comment by eirinnm** [comment]
@dnkl that is good info. I note that in ubuntu's terminal, the default presentation is the colour emoji. I had to use the text selector, /ufe0e, to get the monochrome version.

echo -e 'text presentation: ✍\ufe0e emoji presentation: ✍\ufe0f'

**Comment by jarun** [comment]
Guys, I need a PR with a minimal set.
Also, if possible a link to more emojis in the wiki so a user can build on top of it.

**Comment by N-R-K** [comment]
> @N-R-K or @eirinnm would you like to try?

Hmm, I thought I had already replied to this, but I guess I forgot. Anyways, I'll be a bit busy the next couple days, so don't think I'll be able to make time for this.

**Comment by jarun** [comment]
@N-R-K not in a hurry for this. You can try once you have some time.

**Comment by dkabus** [comment]
I actually got this working two days ago in my build of st independently (i.e. without knowing of this discussion here). I just padded the emoji with spaces depending on how many characters it uses, which might be suboptimal. The result speaks for itself though:

![screenshot_20220503_174201](https://user-images.githubusercontent.com/11159390/167013507-8331211e-1de8-4771-8522-baebc013fa79.png)

After I have merged the list with emoji above with my own, I will do a pull request.
